src/replica/testHttpAsyncReq.cc

```diff
@@ -0,0 +1,546 @@
+        BOOST_CHECK(req->cancel());
+    });
+    cancelReqTimer.start();
+
+    thread serviceThread([&io_service] () {
+        io_service.run();
+    });
+    serviceThread.join();
+}
+
+// Testing request cancelation before starting the request.
+
+BOOST_AUTO_TEST_CASE(HttpAsyncReq_cancelled_before_started) {
+
+    LOGS_INFO("HttpAsyncReq_cancelled_before_started");
+
+    asio::io_service io_service;
+
+    // The deadline timer limits the duration of the test to prevent the test from
+    // being stuck for longer than expected.
+    ::AsyncTimer testAbortTimer(io_service, 300, [] (auto expirationIvalMs) {
+        LOGS_INFO("HttpAsyncReq_cancelled_before_started: test exceeded the time budget of " << expirationIvalMs << " ms");
+        std::exit(1);
+    });
+    testAbortTimer.start();
+
+    // Set up and start the server
+    ::HttpServer httpServer;
+    httpServer.server()->addHandler("GET", "/quick",
+        [] (qhttp::Request::Ptr const& req, qhttp::Response::Ptr const& resp) {
+            resp->sendStatus(200);
+        })
+    ;
+    httpServer.start();
+
+    // Submit a request.
+    string const url = "http://127.0.0.1:" + to_string(httpServer.port()) + "/quick";
+    string const method = "GET";
+    shared_ptr<HttpAsyncReq> const req = HttpAsyncReq::create(
+        io_service,
+        [&testAbortTimer](auto const& req) {
+            testAbortTimer.cancel();
+            BOOST_CHECK(req->state() == HttpAsyncReq::State::CANCELLED);
+        },
+        method, url
+    );
+
+    // Cancel right away.
+    BOOST_CHECK(req->cancel());
+    BOOST_CHECK(req->state() == HttpAsyncReq::State::CANCELLED);
+    BOOST_CHECK(!req->cancel());    // since the request was already cancelled
+
+    // It's not allowed to start the cancelled requests
+    BOOST_CHECK_THROW(req->start(), std::logic_error);
+
+    thread serviceThread([&io_service] () {
+        io_service.run();
+    });
+    serviceThread.join();
+}
+
+// Testing an ability of HttpAsyncReq to wait before the server will start on
+// some fixed port.
+
+BOOST_AUTO_TEST_CASE(HttpAsyncReq_delayed_server_start) {
+
+    LOGS_INFO("HttpAsyncReq_delayed_server_start");
+
+    asio::io_service io_service;
+
+    // The deadline timer limits the duration of the test to prevent the test from
+    // being stuck for longer than expected.
+    ::AsyncTimer testAbortTimer(io_service, 5000, [] (auto expirationIvalMs) {
+        LOGS_INFO("HttpAsyncReq_delayed_server_start: test exceeded the time budget of " << expirationIvalMs << " ms");
+        std::exit(1);
+    });
+    testAbortTimer.start();
+
+    // Set up the server on the fixed port. The server start will be delayed by the timer.
+    // Since we don't have any guarantee that the port is not in use by some other process,
+    // and should this port be not available the timer will cancel the request to allow
+    // this test to pass.
+    uint16_t const fixedPort = 8080;
```

Comment:
Fixed port is not great, but probably okay.  You might want to pick something less commonly used than 8080?

An alternative might be to create a new asio socket here and `bind()` it to an address with port 0.  This will allocate an unused port.  `qhttp` uses SO_REUSEADDR option already on its listen port, so you could leave the socket open and just pass that newly allocated port to the `qhttp::Server` constructor, and I think it would work and avoid the potential collision problem?

Reply:
I will pick some other (large number) port.

As far as your `bind` idea is concerned, I have already explored a kind of similar track by starting and stopping `qhttp` using this series of calls:
```
auto ptr = qhttp::create(io_service, 0);
ptr->start();
ptr->stop();
...
// Then delayed start in the ASYNC timer callback hoping the server would get the same port
{
  ptr->start()
}
```
This all ended up with some really bad crash in ... `XRootD` stack 😸 

I could also modify this algorithm by dropping the first instance of `qhttp` and replacing it with another one. Would give me the same result as your `bind` proposal?